### PR TITLE
Tensor construction: combine Resize+mutable_data - 1/4

### PR DESCRIPTION
**Pull request by jerryzh168**
Summary:
Codemod generated with clangr shard mode, 25 files per diff,
motivation: https://github.com/pytorch/pytorch/pull/12407

Differential Revision: D13007246

### CODE REVIEW

File: caffe2/operators/batch_gather_ops.h [hunk 1]
```diff
@@ -22,15 +22,14 @@ class BatchGatherOp final : public Operator<Context> {
...
-    auto* output = Output(0);
 
     CAFFE_ENFORCE_GE(data.dim(), 2, "DATA should be at least 2-D");
 
     vector<int64_t> shape;
     shape.push_back(data.size(0));
     shape.insert(shape.end(), indices.sizes().begin(), indices.sizes().end());
     shape.insert(shape.end(), data.sizes().begin() + 2, data.sizes().end());
-    output->Resize(shape);
+    auto* output = Output(0, shape, at::dtype<float>());
```
Comment (smessmer) [inline caffe2/operators/batch_gather_ops.h]: The code below doesn't look like this is always float. It only uses float if data.IsType<float>(). It uses data.dtype(), maybe you should use that too?

File: caffe2/operators/cross_entropy_op.cc [hunk 2]
```diff
@@ -78,13 +78,15 @@ bool SigmoidCrossEntropyWithLogitsOp<float, CPUContext>::RunOnDevice() {
...
+  std::vector<int64_t> dims;
   if (logits.dim() == 0) {
-    out->Resize(std::vector<int64_t>{});
+    dims = std::vector<int64_t>{};
   } else {
-    std::vector<int64_t> dims(logits.sizes().begin(), logits.sizes().end() - 1);
-    out->Resize(dims);
+    std::vector<int64_t> dims_(
+        logits.sizes().begin(), logits.sizes().end() - 1);
+    dims = dims_;
```
Comment (smessmer) [inline caffe2/operators/cross_entropy_op.cc]: This code seems overly complicated. Proposing some changes that make it more readable, and also faster (because it avoids the vector copy you have in line 87 and moves instead).

    std::vector<int64_t> dims;
    if (logits.dim() != 0) {
      dims = std::vector<int64_t>(
          logits.sizes().begin(), logits.sizes().end() - 1);
    }

File: caffe2/operators/cross_entropy_op.cc [hunk 3]
```diff
@@ -160,13 +162,16 @@ bool WeightedSigmoidCrossEntropyWithLogitsOp<float, CPUContext>::RunOnDevice() {
...
```
Comment (smessmer) [inline caffe2/operators/cross_entropy_op.cc]: same here

File: caffe2/operators/elementwise_ops.h [hunk 4]
```diff
@@ -147,19 +147,19 @@ class BinaryElementwiseWithArgsOp final : public Operator<Context> {
...
     std::vector<int> B_dims;
+    std::vector<int64_t> C_dims;
 
     if (legacy_broadcast_) {
-      CAFFE_ENFORCE_NE(
-          C,
-          &B,
+      CAFFE_ENFORCE(
+          !IsInputOutputAlias(1, 0),
           "In-place is allowed only with the first tensor when "
```
Comment (smessmer) [inline caffe2/operators/elementwise_ops.h]: this wasn't changed in this diff, but this message is weird. if it is legacy broadcast, then it asserts that it is not inplace, saying "inplace is only allowed if legacy broadcast"? Is this a bug?

File: caffe2/operators/elementwise_ops.h [hunk 5]
```diff
@@ -176,17 +176,19 @@ class BinaryElementwiseWithArgsOp final : public Operator<Context> {
...
+      auto C_dims_int =
           elementwise_ops_utils::ComputeBinaryBroadcastForwardDims(
               A_dims, B_dims);
-      if (C == &A) {
-        CAFFE_ENFORCE_EQ(C_dims, A_dims);
-      } else if (C == &B) {
-        CAFFE_ENFORCE_EQ(C_dims, B_dims);
-      } else {
-        C->Resize(C_dims);
+      std::copy(
```
Comment (smessmer) [inline caffe2/operators/elementwise_ops.h]: m we should probably change the return type of ComputeBinaryBroadcastForwardDims to vector<int64_t>, so this weird copy isn't necessary. Probably shouldn't be done in this diff though.

**Review by smessmer — APPROVED** [review]

File: caffe2/operators/elementwise_ops.h (hunk 4 above)
Comment (jerryzh168) [reply caffe2/operators/elementwise_ops.h]: looks like this is asserting we can only allow Output(0) being alias of Input(0), this is checking Output(0) is not an alias of Input(1). cc @BIT-silence

File: caffe2/operators/elementwise_ops.h (hunk 5 above)
Comment (jerryzh168) [reply caffe2/operators/elementwise_ops.h]: yeah, there are some cases in "caffe2/utils/math.h" as well
